src/main/java/org/quiltmc/loader/impl/launch/knot/KnotClassDelegate.java

```diff
@@ -129,6 +132,80 @@ private IMixinTransformer getMixinTransformer() {
 		return mixinTransformer;
 	}
 
+	Class<?> loadClass(String name, ClassLoader parent, boolean resolve) throws ClassNotFoundException {
+		Class<?> c = loadClassOnly(name, parent);
+		if (resolve) {
+			itf.resolveClassFwd(c);
+		}
+		return c;
+	}
+
+	private Class<?> loadClassOnly(String name, ClassLoader parent) throws ClassNotFoundException {
+		Class<?> c = itf.findLoadedClassFwd(name);
+		if (c != null) {
+			return c;
+		}
+
+		try {
+			c = tryLoadClass(name, false);
+		} catch (IllegalQuiltInternalAccessError e) {
+			// This happens when loading a class that needs a quilt-loader class to load:
+			// for example if it extends or implements one,
+			// or if the verifier needs to load one to check up-casts.
+			IllegalQuiltInternalAccessError e2 = new IllegalQuiltInternalAccessError("Failed to load the class " + name + "!");
+			e2.initCause(e);
+			throw e2;
+		}
+
+		if (c != null) {
+			return c;
+		}
+
+		c = parent.loadClass(name);
+
+		if (c != null) {
+			QuiltLoaderInternal internal = c.getAnnotation(QuiltLoaderInternal.class);
+			QuiltLoaderInternalType type;
+			if (internal != null) {
+				type = internal.value();
+			} else if (name.startsWith("org.quiltmc.loader.impl")) {
+				type = QuiltLoaderInternalType.LEGACY_EXPOSED;
+				Log.warn(LogCategory.GENERAL, c + " isn't annotated with @QuiltLoaderInternal!");
+			} else if (name.startsWith("org.quiltmc.loader.api.plugin")) {
+				type = QuiltLoaderInternalType.PLUGIN_API;
+				Log.warn(LogCategory.GENERAL, c + " isn't annotated with @QuiltLoaderInternal!");
+			} else {
+				return c;
+			}
+
+			switch (type) {
+				case LEGACY_EXPOSED: {
+					// TODO: Disable this when we can generate a report with this information!
+					Log.warn(
+						LogCategory.GENERAL, "Found access to quilt-loader internal api " + name
+							+ " - please ask the quilt-loader project to add an api for this, to ensure this functionality won't break in a future update!",
+						new Throwable()
+					);
+					break;
+				}
+				case NEW_INTERNAL:
+				default: {
+					throw new IllegalQuiltInternalAccessError("! Quilt-loader internal " + c
+						+ "\nPlease don't use this, instead ask us to declare a new public API that we can guarantee backwards compatibility for!");
+				}
+				case PLUGIN_API: {
+					throw new IllegalQuiltInternalAccessError("! Quilt-loader plugin-only internal " + c
+						+ "\nPlease don't use this, instead ask us to declare a new public API that we can guarantee backwards compatibility for!");
+				}
```

Comment:
Something else to consider is that often internal API does have a public equivalent, but they just missed it.

Reply:
That's a good point - perhaps the annotation could include a description of what might be a replacement?

Reply:
On second thought, perhaps a separate annotation that could be applied to methods and fields too, to be checked in... something else?